src/compiler/types.ts

```diff
@@ -832,6 +833,13 @@ namespace ts {
         expression: Expression;
     }
 
+    export interface PrivateName extends PrimaryExpression, Declaration {

```

Comment:
Why is `PrivateName` a `PrimaryExpression` and a `Declaration`? I don't believe it should be classified as either since it can't be used on its own as an expression. It should probably just extend `Node`, just like `ComputedPropertyName`.

Reply:
That's probably to copy the definition of `Identifier`, which is a Declaration because of Javascript. In the second line of:

```js
var NS = {}
NS.f = x => x + 1
```

`NS` is an identifier that is a secondary declaration of `NS`. This doesn't apply to private names because they will never be treated as expando, since they come from a property declaration.

Reply:
My initial reaction is that you don't need a separate data structure, since private names aren't very different from normal identifiers, unlike ComputedPropertyNames. An identifier with originalKeywordKind === SyntaxKind.PrivateName is good enough to distinguish private names when necessary. That'll allow you to reuse all the identifier machinery, in the parser and language service especially.

I'll keep reading to see whether there are any advantages to having the two unique.

Reply:
The trade-offs (in my limited knowledge) seem to be:

### Option 1: Reuse `SyntaxKind.Identifier` for private names

#### Advantages

- Way less code
- Can make good, specific error messages in the binder and checker

#### Disadvantages

- The AST will be less like the AST in the spec text. This could be especially difficult for users of the compiler API, who might expect to see `PrivateName` nodes.
- We'll have to hunt down all the places where private names *shouldn't* be, and error in each one of them–this may be a bug vector

#### Implementation

Here's an implementation: https://github.com/mheiber/TypeScript/commit/c2d182158339534ad787ab86422b72631836979f

I used a new `TokenFlag`, but your suggestion of using `originalKeywordKind` would be better (less waste of the bit budget).

### Option 2: `SyntaxKind.PrivateName`

#### Advantages

- maintainability
- AST shape is like that in the spec-text

#### Disadvantages

- The error message for `const #foo = 3` is pretty bad

### Option 3: loose parsing for `SyntaxKind.PrivateName` 

- In this implementation, there would be separate `Identifier` and `PrivateName` data structures.
- The main difference from the current approach (Option 2) is that we would allow `PrivateName` nodes lots of places in the AST where private names are not actually allowed in JS, such as in `const #foo = 3`.
- Then we'd catch these bad usages in the checker.

This seems like it would have the biggest cost to the codebase (LOC and maintenance of places where private names are allowed) but perhaps the best user experience (nice error messages and spec-like AST).

Reply:
Resolving: we'd decided to keep option 2 for now (SyntaxKind.PrivateName) but plan to look into improving error messages for `const #foo = 3` outside this pr